### PR TITLE
Add `MemoryRecordStore`

### PR DESCRIPTION
**Pull request by desertaxle**
This PR adds a `MemoryRecordStore` implementation for testing transaction locking. The implementation uses `threading.Lock` to handle record locking. This implementation doesn't exactly match the `RecordStore` base class because it accepts a `BaseResult` instead of an arbitrary value. The assumption is that a result will be created before creating a `TransactionRecord`. If that's a valid assumption, I can update the `RecordStore.write` signature and the integration with `Transaction` in a subsequent PR.

Related to #14914 

### Checklist

- [x] This pull request references any related issue by including "closes `<link to issue>`"
  - If no issue exists and your change is not a small fix, please [create an issue](https://github.com/PrefectHQ/prefect/issues/new/choose) first.
- [x] If this pull request adds new functionality, it includes unit tests that cover the changes
- [ ] If this pull request removes docs files, it includes redirect settings in `mint.json`.
- [ ] If this pull request adds functions or classes, it includes helpful docstrings.

### CODE REVIEW

File: src/prefect/records/in_memory.py [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+
+    holder: str
+    lock: threading.Lock
+    expiration_timer: Optional[threading.Timer]
+
+
+_locks: Dict[str, _LockInfo] = {}
+
+
+class InMemoryTransactionRecord(TransactionRecord):
```
Comment (cicdw) [inline src/prefect/records/in_memory.py:27]: slight preference for `MemoryTransactionRecord` instead if `InMemory`

**Review by cicdw — COMMENTED** [review]
minor request for a name change

File: src/prefect/records/base.py [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+from prefect.results import BaseResult
+
+
+class RecordStore:
+    def read(self, key: str):
+        raise NotImplementedError
+
+    def write(self, key: str, value: dict):
+        raise NotImplementedError
```
Comment (chrisguidry) [inline src/prefect/records/base.py:16]: What are the return types on these?

File: src/prefect/records/base.py [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+import socket
+import threading
+from contextlib import contextmanager
+from dataclasses import dataclass
+from typing import Optional
+
+from prefect.results import BaseResult
+
+
+class RecordStore:
```
Comment (chrisguidry) [inline src/prefect/records/base.py:11]: Maybe use `abc` to signal what methods are abstract?

File: src/prefect/records/in_memory.py [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+                    expiration_timer = threading.Timer(
+                        hold_timeout, self._expire_lock, args=(key,)
+                    )
+                    expiration_timer.start()
+                self._locks[key] = _LockInfo(
+                    holder=holder, lock=lock, expiration_timer=expiration_timer
+                )
+                return True
+            elif self._locks[key]["holder"] == holder:
+                return True
```
Comment (chrisguidry) [inline src/prefect/records/in_memory.py:91]: :heart:

File: tests/records/in_memory.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+
+from prefect.records.in_memory import MemoryRecordStore
+from prefect.results import ResultFactory
+
+
+class TestInMemoryRecordStore:
+    def test_singleton(self):
+        store1 = MemoryRecordStore()
+        store2 = MemoryRecordStore()
+        assert store1 is store2
```
Comment (chrisguidry) [inline tests/records/in_memory.py:15]: :fire:

**Review by chrisguidry — APPROVED** [review]
The tests are lovely!

File: src/prefect/records/base.py (hunk 2 above)
Comment (desertaxle) [reply src/prefect/records/base.py:16]: The signatures of the `MemoryRecordStore` and the `ResultFactoryStore` don't align right now, so I'm leaving the typing on the existing methods as is for now. I plan to update the typing/signature for `RecordStore` and `ResultFactoryStore` when I get `Transaction` working with `MemoryRecordStore` in the next PR.

File: src/prefect/records/base.py (hunk 3 above)
Comment (desertaxle) [reply src/prefect/records/base.py:11]: I can do this in the next PR when all the signatures are aligned.

File: src/prefect/records/base.py (hunk 2 above)
Comment (chrisguidry) [reply src/prefect/records/base.py:16]: Sounds great!

**Review by cicdw — CHANGES_REQUESTED** [review]
LGTM! For consistency can you change the names of the test file and the record file to `memory.py` from `in_memory.py`?

**Review by cicdw — APPROVED** [review]

**Pull request merged**